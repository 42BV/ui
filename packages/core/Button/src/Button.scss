@import '~@42.nl/ui-core-styling/dist/scss/abstracts/variables';

.Button {
  .btn {
    font-size: 0.9rem;
    position: relative;
    overflow: hidden;

    @each $color, $value in $theme-colors {
      &.btn-#{$color} {
        &:hover,
        &:active:hover {
          background-color: darken($value, 6%);
        }
      }
    }

    .material-icons {
      vertical-align: middle;
      font-size: 1rem;
      margin-bottom: 2px;
    }

    .Spinner {
      position: relative;
      top: -2px;
      left: -5px;
    }
  }

  @each $color, $value in $theme-colors {
    &.#{$color} {
      .Spinner {
        stroke: $value;
      }
    }
  }
}
