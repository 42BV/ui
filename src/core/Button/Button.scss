.button-spinner {
  stroke: currentColor;
}

.btn-outline-info {
  color: var(--bs-text-info-contrast);
}

.btn-outline-warning {
  color: var(--bs-text-warning-contrast);
}

.button-icon {
  font-size: 1.1em;
}
