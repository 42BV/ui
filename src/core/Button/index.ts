export { Button, Props, ButtonSize, ButtonIconPosition } from './Button';
