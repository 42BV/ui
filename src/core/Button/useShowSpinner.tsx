import { useEffect, useState } from 'react';

/**
 * Only show the Spinner when 200 milliseconds have past.
 * This prevents very short loading periods from showing the animation,
 * seeing the Spinner very briefly looks very glitchy.
 *
 * @param show
 */
export function useShowSpinner(show: boolean): boolean {
  const [ showSpinner, setShowSpinner ] = useState(show);

  useEffect(() => {
    let timer: number | undefined = undefined;
    if (show) {
      timer = window.setTimeout(() => {
        setShowSpinner(true);
      }, 200);
    } else {
      setShowSpinner(false);
    }

    return () => {
      window.clearTimeout(timer);
    };
  }, [ show ]);

  return showSpinner;
}
