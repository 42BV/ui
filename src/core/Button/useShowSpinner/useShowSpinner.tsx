import { useEffect, useState } from "react";

/**
 * Only show the Spinner when 200 milliseconds have past.
 * This prevents very short loading periods from showing the animation,
 * seeing the Spinner very briefly looks very glitchy.
 *
 * @param show
 */
export function useShowSpinner(show: boolean): boolean {
  const [showSpinner, setShowSpinner] = useState(show);

  useEffect(() => {
    if (show) {
      const timer = window.setTimeout(() => {
        setShowSpinner(true);
      }, 200);

      return () => {
        window.clearTimeout(timer);
      };
    } else {
      setShowSpinner(false);
    }
  }, [show]);

  return showSpinner;
}
