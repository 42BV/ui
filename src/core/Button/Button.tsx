import React from 'react';
import { Button as RSButton } from 'reactstrap';

import Spinner from '../Spinner/Spinner';
import { Icon, IconType } from '../Icon';
import useShowSpinner from './useShowSpinner';

import { Color } from '../types';

export type ButtonSize = 'sm' | 'md' | 'lg';

export type IconPosition = 'left' | 'right';

export interface BaseProps {
  /**
   * Optionally the type of button it is, defaults to 'button'.
   *
   * @default button
   */
  type?: 'button' | 'submit' | 'reset';

  /**
   * Optionally the color of the button.
   */
  color?: Color;

  /**
   *  Optional callback for what needs to happen when the button is clicked.
   */
  onClick?: (event: React.MouseEvent<HTMLElement>) => any;

  /**
   * Whether or not the action you are performing is currently in
   * progress. If so a spinner is rendered inside of the button.
   * This behavior is optional and default to `false`.
   *
   * @default false
   */
  inProgress?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optionally whether the button is disabled
   *
   * Defaults to `false`
   */
  disabled?: boolean;
}

interface WithIconAndText extends BaseProps {
  /**
   * Optionally the Icon you want to use.
   */
  icon: IconType;

  iconPosition?: IconPosition;

  /**
   * Optionally the size of the icon in pixels.
   */
  iconSize?: number;

  /**
   * Optionally the text of the button.
   */
  children: React.ReactNode;

  /**
   * Optionally whether or not to show the button only as an outline.
   */
  outline?: boolean;

  /**
   * Optionally the size of the button.
   *
   * Defaults to 'md'.
   */
  size?: ButtonSize;

  /**
   * Optionally whether or not the button should take the full width
   * available.
   *
   * Defauts to `false`
   */
  fullWidth?: boolean;
}

interface WithIcon extends BaseProps {
  /**
   * The Icon you want to use.
   */
  icon: IconType;

  iconPosition?: IconPosition;

  /**
   * Optionally the size of the icon in pixels.
   */
  iconSize?: number;

  children?: never;
  outline?: never;
  size?: never;
  fullWidth?: never;
}

interface WithText extends BaseProps {
  /**
   * Optionally the text of the button.
   */
  children: React.ReactNode;

  /**
   * Optionally whether or not to show the button only as an outline.
   */
  outline?: boolean;

  /**
   * Optionally the size of the button.
   *
   * Defaults to 'md'.
   */
  size?: ButtonSize;

  /**
   * Optionally whether or not the button should take the full width
   * available.
   *
   * Defauts to `false`
   */
  fullWidth?: boolean;

  icon?: never;
  iconPosition?: never;
}

export type Props = WithIcon | WithText | WithIconAndText;

/**
 * The Button component is a clickable element which can
 * either be shown as an Icon or as a ReactStrap Button.
 *
 * The Button component also has a notion of being `inProgress` when it is
 * it shows a Spinner. But only when `inProgress` has been true for more
 * than 200 milliseconds. Useful for showing that a button is indeed
 * clicked when performing some call to the back-end.
 */
export default function Button({
  type = 'button',
  color = 'primary',
  inProgress,
  className = '',
  onClick,
  ...props
}: Props) {
  const showSpinner = useShowSpinner(!!inProgress);

  function handleOnClick(event: React.MouseEvent<HTMLElement>) {
    if (!onClick) {
      return;
    }

    if (!inProgress) {
      onClick(event);
    }
  }

  const disabled = 'disabled' in props ? props.disabled : undefined;

  if ('children' in props) {
    const children = props.children;
    const outline = 'outline' in props ? props.outline : undefined;
    const size = 'size' in props ? props.size : 'md';
    const fullWidth = 'fullWidth' in props ? props.fullWidth : false;

    const buttonProps = {
      type,
      size,
      color,
      outline,
      block: fullWidth
    };

    return (
      <span className={`button ${className} ${color}`}>
        <RSButton
          onClick={handleOnClick}
          disabled={inProgress || disabled}
          {...buttonProps}
        >
          {showSpinner ? (
            <Spinner size={16} color={outline ? '' : 'white'} />
          ) : 'icon' in props && props.icon ? (
            <Icon
              icon={props.icon}
              className={
                'material-icons-' +
                ('iconPosition' in props ? props.iconPosition : 'left')
              }
              size={props.iconSize}
            />
          ) : null}
          {children}
        </RSButton>
      </span>
    );
  } else {
    // We know that at this point it must be have icon,
    // because the Button now extends WithIcon.
    const icon = props.icon;
    const iconCast = icon as IconType;

    return (
      <span className={`button ${className} ${color}`}>
        {showSpinner ? (
          // Color is empty string so we can override the color
          <Spinner size={24} color="" />
        ) : (
          <Icon
            onClick={handleOnClick}
            icon={iconCast}
            color={color}
            disabled={inProgress || disabled}
            size={props.iconSize}
          />
        )}
      </span>
    );
  }
}

export function isWithIcon(props: Props): props is WithIcon {
  return props.children === undefined && props.icon !== undefined;
}

export function isWithIconAndText(props: Props): props is WithIconAndText {
  return props.children !== undefined && props.icon !== undefined;
}

export function isWithText(props: Props): props is WithText {
  return props.children !== undefined && props.icon === undefined;
}
