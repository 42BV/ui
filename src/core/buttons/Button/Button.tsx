import React from 'react';
import { Button as RSButton } from 'reactstrap';

import { Spinner } from '../../Spinner/Spinner';
import { Icon, IconType } from '../../Icon';
import { useShowSpinner } from '../useShowSpinner';

import {
  ButtonIconPosition,
  ButtonProps as GButtonProps,
  ButtonSize
} from '../../types';
import classNames from 'classnames';

type ButtonProps = {
  children: React.ReactNode;
} & Partial<GButtonProps>;

function getWidget(
  showSpinner: boolean,
  size: ButtonSize,
  outline: boolean,
  iconPosition: ButtonIconPosition,
  icon?: IconType
) {
  if (showSpinner) {
    return (
      <Spinner
        size={getSpinnerSize(size)}
        color={outline ? '' : 'white'}
        className={iconPosition === 'left' ? 'me-2' : 'ms-2'}
      />
    );
  } else if (icon) {
    return (
      <Icon
        icon={icon}
        className={classNames(
          'button-icon',
          iconPosition === 'left' ? 'me-2' : 'ms-2'
        )}
      />
    );
  }
  return null;
}

/**
 * The buttons component is a clickable element which can
 * either be shown as an Icon or as a ReactStrap buttons.
 *
 * The buttons component also has a notion of being `inProgress` when it is
 * it shows a Spinner. But only when `inProgress` has been true for more
 * than 200 milliseconds. Useful for showing that a button is indeed
 * clicked when performing some call to the back-end.
 */
export function Button({
  type = 'button',
  color = 'primary',
  inProgress,
  className = '',
  onClick,
  size = 'md',
  outline,
  fullWidth,
  iconPosition = 'left',
  icon,
  disabled,
  children
}: ButtonProps) {
  const showSpinner = useShowSpinner(!!inProgress);

  function handleOnClick(event: React.MouseEvent<HTMLElement>) {
    if (!onClick) {
      return;
    }
    if (!inProgress) {
      onClick(event);
    }
  }

  const buttonProps = {
    type,
    size,
    color,
    outline,
    block: fullWidth
  };

  const widget = getWidget(showSpinner, size, !!outline, iconPosition, icon);

  return (
    <span
      className={classNames(
        'button',
        fullWidth ? 'd-block' : 'd-inline-block',
        className,
        color
      )}
    >
      <RSButton
        onClick={handleOnClick}
        disabled={inProgress || disabled}
        {...buttonProps}
      >
        <div className="d-flex justify-content-center align-items-center">
          {iconPosition === 'left' && widget}
          {children}
          {iconPosition === 'right' && widget}
        </div>
      </RSButton>
    </span>
  );
}

// Based on the sizes in pixels of a button with text but without an icon
export function getSpinnerSize(size?: ButtonSize): number {
  switch (size) {
    case 'lg':
      return 19;

    case 'md':
      return 16;

    case 'sm':
      return 12;
  }
  return 7;
}
