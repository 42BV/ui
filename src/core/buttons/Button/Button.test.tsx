import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { Button, getSpinnerSize } from './Button';
import { useShowSpinner } from '../useShowSpinner';

describe('Component: buttons', () => {
  describe('ui', () => {
    describe('button', () => {
      describe('normal', () => {
        test('inProgress is true', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={true} iconPosition="right">
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('inProgress is false', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={false}>
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });

      describe('outline', () => {
        test('inProgress is true', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={true} outline={true}>
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('inProgress is false', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={false} outline={true}>
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });

      describe('disabled', () => {
        test('is disabled', () => {
          const { container } = render(
            <Button onClick={jest.fn()} disabled={true}>
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('is enabled', () => {
          const { container } = render(
            <Button onClick={jest.fn()} disabled={false}>
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });

      describe('size', () => {
        test('use md by default', () => {
          const { container } = render(
            <Button onClick={jest.fn()}>Save</Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('can override size', () => {
          const { container } = render(
            <Button onClick={jest.fn()} size="sm">
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });

      test('full-width', () => {
        const { container } = render(
          <Button onClick={jest.fn()} fullWidth={true}>
            Save
          </Button>
        );

        expect(container).toMatchSnapshot();
      });
    });

    describe('button with icon', () => {
      describe('normal', () => {
        test('inProgress is true', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={true} icon="save">
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('inProgress is false', () => {
          const { container } = render(
            <Button onClick={jest.fn()} inProgress={false} icon="save">
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('with icon on the right', () => {
          const { container } = render(
            <Button onClick={jest.fn()} icon="save" iconPosition="right">
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });

      describe('outline', () => {
        test('inProgress is true', () => {
          const { container } = render(
            <Button
              onClick={jest.fn()}
              inProgress={true}
              outline={true}
              icon="save"
            >
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });

        test('inProgress is false', () => {
          const { container } = render(
            <Button
              onClick={jest.fn()}
              inProgress={false}
              outline={true}
              icon="save"
            >
              Save
            </Button>
          );

          expect(container).toMatchSnapshot();
        });
      });
    });
  });

  describe('events', () => {
    describe('onClick', () => {
      function setup({
        inProgress,
        hasOnClick
      }: {
        inProgress: boolean;
        hasOnClick: boolean;
      }) {
        const onClickSpy = jest.fn();

        jest
          .spyOn({ useShowSpinner }, 'useShowSpinner')
          .mockReturnValue(inProgress);

        const props = {
          onClick: hasOnClick ? onClickSpy : undefined,
          inProgress
        };

        const { container } = render(<Button {...props}>Save</Button>);

        return { container, onClickSpy };
      }

      describe('button', () => {
        it('should call the onClick callback when inProgress is false', () => {
          const { onClickSpy } = setup({
            inProgress: false,
            hasOnClick: true
          });

          fireEvent.click(screen.getByText('Save'));

          expect(onClickSpy).toHaveBeenCalledTimes(1);
        });

        it('should not call the onClick callback when inProgress is true and ignore the click', () => {
          const { onClickSpy } = setup({
            inProgress: true,
            hasOnClick: true
          });

          fireEvent.click(screen.getByText('Save'));

          expect(onClickSpy).toHaveBeenCalledTimes(0);
        });

        it('should not call the onClick callback when onClick is undefined', () => {
          const { onClickSpy } = setup({
            inProgress: false,
            hasOnClick: false
          });

          fireEvent.click(screen.getByText('Save'));

          expect(onClickSpy).toHaveBeenCalledTimes(0);
        });
      });
    });
  });
});

test('Util: getSpinnerSize', () => {
  expect(getSpinnerSize('lg')).toBe(19);
  expect(getSpinnerSize('md')).toBe(16);
  expect(getSpinnerSize('sm')).toBe(12);
});
