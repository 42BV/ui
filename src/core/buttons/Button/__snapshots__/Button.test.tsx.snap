// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button ui button disabled is disabled 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button disabled is enabled 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button full-width 1`] = `
<div>
  <span
    class="button d-block primary"
  >
    <button
      class="btn btn-primary btn-md d-block w-100"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button normal inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button normal inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
        <svg
          class="spinner ms-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="white"
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button outline inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button outline inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke=""
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button size can override size 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-sm"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button size use md by default 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button with icon normal inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <i
          class="icon button-icon me-2 material-icons"
        >
          save
        </i>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button with icon normal inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="white"
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button with icon normal with icon on the right 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
        <i
          class="icon button-icon ms-2 material-icons"
        >
          save
        </i>
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button with icon outline inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <i
          class="icon button-icon me-2 material-icons"
        >
          save
        </i>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui button with icon outline inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke=""
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: Button ui icon only disabled is disabled 1`] = `
<div>
  <span
    class="button primary d-inline-block"
  >
    <i
      class="icon material-icons text-primary icon--disabled"
      style="font-size: 24px;"
    >
      save
    </i>
  </span>
</div>
`;

exports[`Component: Button ui icon only disabled is enabled 1`] = `
<div>
  <span
    class="button primary d-inline-block"
  >
    <i
      class="icon material-icons text-primary clickable"
      style="font-size: 24px;"
    >
      save
    </i>
  </span>
</div>
`;

exports[`Component: Button ui icon only full-width full width and icon left 1`] = `
<div>
  <span
    class="button primary d-flex justify-content-start"
  >
    <i
      class="icon material-icons text-primary clickable"
      style="font-size: 24px;"
    >
      save
    </i>
  </span>
</div>
`;

exports[`Component: Button ui icon only full-width full width and icon right 1`] = `
<div>
  <span
    class="button primary d-flex justify-content-end"
  >
    <i
      class="icon material-icons text-primary clickable"
      style="font-size: 24px;"
    >
      save
    </i>
  </span>
</div>
`;

exports[`Component: Button ui icon only inProgress inProgress is false 1`] = `
<div>
  <span
    class="button primary d-inline-block"
  >
    <i
      class="icon material-icons text-primary clickable"
      style="font-size: 24px;"
    >
      save
    </i>
  </span>
</div>
`;

exports[`Component: Button ui icon only inProgress inProgress is true 1`] = `
<div>
  <span
    class="button primary d-inline-block"
  >
    <svg
      class="spinner"
      height="24"
      viewBox="25 25 50 50"
      width="24"
    >
      <circle
        class="path"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke=""
        stroke-miterlimit="10"
        stroke-width="4"
      />
    </svg>
  </span>
</div>
`;

exports[`Component: Button ui icon only that when no icon is provided it will fallback to block 1`] = `
<div>
  <span
    class="button primary d-inline-block"
  >
    <i
      class="icon material-icons text-primary clickable"
      style="font-size: 24px;"
    >
      block
    </i>
  </span>
</div>
`;

exports[`Component: buttons ui button disabled is disabled 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button disabled is enabled 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button full-width 1`] = `
<div>
  <span
    class="button d-block primary"
  >
    <button
      class="btn btn-primary btn-md d-block w-100"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button normal inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button normal inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
        <svg
          class="spinner ms-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="white"
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button outline inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button outline inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke=""
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button size can override size 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-sm"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button size use md by default 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button with icon normal inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <i
          class="icon button-icon me-2 material-icons"
        >
          save
        </i>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button with icon normal inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke="white"
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button with icon normal with icon on the right 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        Save
        <i
          class="icon button-icon ms-2 material-icons"
        >
          save
        </i>
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button with icon outline inProgress is false 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md"
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <i
          class="icon button-icon me-2 material-icons"
        >
          save
        </i>
        Save
      </div>
    </button>
  </span>
</div>
`;

exports[`Component: buttons ui button with icon outline inProgress is true 1`] = `
<div>
  <span
    class="button d-inline-block primary"
  >
    <button
      class="btn btn-outline-primary btn-md disabled"
      disabled=""
      type="button"
    >
      <div
        class="d-flex justify-content-center align-items-center"
      >
        <svg
          class="spinner me-2"
          height="16"
          viewBox="25 25 50 50"
          width="16"
        >
          <circle
            class="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            stroke=""
            stroke-miterlimit="10"
            stroke-width="4"
          />
        </svg>
        Save
      </div>
    </button>
  </span>
</div>
`;
